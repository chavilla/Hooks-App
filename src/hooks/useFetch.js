import { useEffect, useRef, useState } from "react";

const useFecth = (url, loading) => {
  const isMounted = useRef(true);
  const [state, setState] = useState({ data: null, loading, error: null });

  useEffect(() => {
    // cleanup
    return () => {
      isMounted.current = false;
    };
  }, []);

  useEffect(() => {
    setState({ data: null, loading: true, error: null });
    fetch(url)
      .then((response) => response.json())
      .then((data) => {
        if (isMounted) {
          setState({
            data,
            loading: false,
            error: null,
          });
        }
      });
  }, [url]);

  return {
    state,
  };
};

export default useFecth;
