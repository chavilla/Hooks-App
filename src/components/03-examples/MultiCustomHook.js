import React from "react";
import { useCounter } from "../../hooks/useCounter";
import useFecth from "../../hooks/useFetch";

const MultiCustomHook = () => {
  const { counter, increment } = useCounter();

  const {
    data, loading 
  } = useFecth(`https://www.breakingbadapi.com/api/quotes/${counter}`, true);

  // Condicional Importante
  const { author, quote } = !!data && data[0];

  //console.log(loading);

  return (
    <div>
      <p className='counter'>{counter}</p>
      <h1>Breaking Quotes</h1>
      {loading ? (
        <div className="alert alert-info text-center">Loading...</div>
      ) : (
        <blockquote className="blockquote text-right">
          <p className="mb-0">{quote}</p>
          <footer className="block">{author}</footer>
        </blockquote>
      )}
      <button className="btn btn-primary" onClick={increment}>
        Siguiente quote
      </button>
    </div>
  );
};

export default MultiCustomHook;
