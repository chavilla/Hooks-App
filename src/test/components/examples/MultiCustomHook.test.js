import { shallow } from 'enzyme';
import React from 'react';
import MultiCustomHook from '../../../components/03-examples/MultiCustomHook';
import { useCounter } from '../../../hooks/useCounter';
import useFecth from '../../../hooks/useFetch';
jest.mock('../../../hooks/useFetch');
jest.mock('../../../hooks/useCounter');

describe('MultiCustomHooks', () => {

    
    beforeEach(()=>{
        useCounter.mockReturnValue({
            counter: 10,
            increment: () => {}
        });
    });


    test('should to create a snapshot', () => {

        useFecth.mockReturnValue({
            data: null, loading: true, error: null 
        });

        const wrapper=shallow(<MultiCustomHook />);

        expect(wrapper).toMatchSnapshot();
    });

    test('should to show the info', () => {

        useFecth.mockReturnValue({
            data: [
                { 
                    author: 'Juan',
                    quote:'Hola a todos'
                }
            ], loading: false, error: null 
        });

        const wrapper=shallow(<MultiCustomHook />);

        console.log(wrapper.find('.counter').text());

        expect(wrapper.find('.alert').exists()).toBe(false);
        expect(wrapper.find('.mb-0').text()).toBe('Hola a todos');
        expect(wrapper.find('footer').text()).toBe('Juan');

    });
})