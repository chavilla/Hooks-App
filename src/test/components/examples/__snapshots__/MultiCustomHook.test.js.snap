// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiCustomHooks should to create a snapshot 1`] = `
<div>
  <p
    className="counter"
  >
    10
  </p>
  <h1>
    Breaking Quotes
  </h1>
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Siguiente quote
  </button>
</div>
`;
